src/context/AppContext.tsx

```diff
@@ -0,0 +1,128 @@
+import {
+  ReactNode,
+  createContext,
+  useEffect,
+  useMemo,
+  useReducer,
+} from 'react';
+import { State } from '../types/State';
+import { Action } from '../types/Action';
+import { AppContextType } from '../types/AppContextType';
+import { getTodos } from '../api/todos';
+import { Error } from '../types/Error';
+import { updateErrorState } from '../utils/errorHandler';
+
+const initialErrorState: Error[] = [
+  { type: 'TodoLoadError', textError: 'Unable to load todos', value: false },
+  {
+    type: 'EmptyTitleError',
+    textError: 'Title should not be empty',
+    value: false,
+  },
+  { type: 'AddTodoError', textError: 'Unable to add a todo', value: false },
+  {
+    type: 'DeleteTodoError',
+    textError: 'Unable to delete a todo',
+    value: false,
+  },
+  {
+    type: 'UpdateTodoError',
+    textError: 'Unable to update a todo',
+    value: false,
+  },
+];
+
+const initialState: State = {
+  todos: [],
+  filteredTodos: [],
+  filter: 'All',
+  errors: initialErrorState,
+  targetTodo: 0, //object with inProgress boolean?
+};
+
+const initialAppContext: AppContextType = {
+  state: initialState,
+  dispatch: () => null,
+};
+
+const reducer = (state: State, action: Action): State => {
+  switch (action.type) {
+    case 'LOAD_TODOS_FROM_SERVER':
+      return {
+        ...state,
+        todos: action.payload,
+        filteredTodos: action.payload,
+      };
+    case 'UPDATE_ERROR_STATUS':
+      return {
+        ...state,
+        errors: updateErrorState(state.errors, action.payload.type),
+      };
+    case 'CLEAR_ERRORS':
+      return {
+        ...state,
+        errors: state.errors.map(error => ({ ...error, value: false })),
+      };
+    case 'ADD_TODO':
+      return {
+        ...state,
+        todos: [...state.todos, action.payload],
+      };
+    case 'SET_TARGET_TODO':
+      return {
+        ...state,
+        targetTodo: action.payload,
+      };
+    case 'SET_FILTER':
+      const filteredTodos = state.todos.filter(todo => {
+        switch (action.payload) {
+          case 'All':
+            return true;
+          case 'Active':
+            return !todo.completed;
+          case 'Completed':
+            return todo.completed;
```

Comment:
Create a enum for `'Active', 'Completed', 'All'` and use it everywhere
```suggestion
          case 'Active':
            return !todo.completed;
            
          case 'Completed':
            return todo.completed;
            
```